wagit/sources/Repository.cpp

```diff
@@ -0,0 +1,75 @@
+//
+// Created by Ranveer Aggarwal on 27/08/20.
+//
+
+#include "../headers/Repository.h"
+
+namespace WAGit
+{
+    Repository::Repository(const Path &path, bool force)
+    {
+        this->workTree = path;
+        this->gitDirectory = path / ".git";
+
+        if (!(force || FS::is_directory(this->gitDirectory)))
+            throw Exceptions::NotAGitRepositoryException(this->gitDirectory.u8string());
+
+        Path configFile = GetRepoPath({"config"});
+        if (!force && !FS::exists(configFile))
+            throw Exceptions::GeneralException("Configuration file missing");
+
+        this->config.InitConfig(configFile);
+
+        if (!force && this->config.GetVersion() != 0)
+            throw Exceptions::GeneralException("Repository Version Not Supported");
+    }
+
+    Path Repository::GetRepoPath(const std::vector<String> &pathToFile)
+    {
+        Path path(this->gitDirectory);
+        for (const auto& subPath: pathToFile)
+        {
+             path /= subPath;
+        }
+
+        return path;
+    }
+
+    bool Repository::GetOrCreateRepoFile(const std::vector<String> &filePath, Path &path, bool mkdir) {
```

Comment:
Where are we creating a file in this if mkdir is true?

Reply:
We don't create a file, we just create corresponding directories leading up to it and return a path object. The name is just wrong. Need to fix it.

Reply:
Then how is this different from the function below it? There also we're just creating the directories and returning the path right? If we're not actually creating the file, I don't understand what the point of this one is.

Reply:
Just for ease of use. Think of it as a wrapper over the `GetRepoPath` function.

Reply:
But yes, optional. Can be removed in the future.